### ISSUE TITLE
WPF: CleanupTemporaryTargetAssembly step after MarkupCompilePass2 intermittently fails to delete target assembly

### ISSUE THREAD

**Issue by vsfeedback** [dotnet/wpf#5255]
_This issue has been moved from [a ticket on Developer Community](https://developercommunity.visualstudio.com/t/WPF:-CleanupTemporaryTargetAssembly-step/1447628)._

---
[severity:Multiple people on my team are impacted (internal MSFT only)]
Some WPF projects require a second pass compile, in which a temporary project is generated and then cleaned up. Sometimes, during this compilation, the "CleanupTemporaryTargetAssembly" step fails to perform cleanup successfully due to the target assembly being locked.

Sample error:
```
MarkupCompilePass2:
  MarkupCompilePass2 successfully generated BAML or source code files.
CleanupTemporaryTargetAssembly:
  Deleting file "Debug\AnyCPU\MyWpfAssembly.dll".
##[error]C:\Windows\Microsoft.NET\Framework\v4.0.30319\Microsoft.WinFx.targets(480,10): Error MSB3061: Unable to delete file "Debug\AnyCPU\MyWpfAssembly.dll". The process cannot access the file 'C:\w\4\s\src\MyWpfAssembly\Debug\AnyCPU\MyWpfAssembly.dll' because it is being used by another process.
C:\Windows\Microsoft.NET\Framework\v4.0.30319\Microsoft.WinFx.targets(480,10): error MSB3061: Unable to delete file "Debug\AnyCPU\MyWpfAssembly.dll". The process cannot access the file 'C:\w\4\s\src\MyWpfAssembly\Debug\AnyCPU\MyWpfAssembly.dll' because it is being used by another process. [C:\w\4\s\src\MyWpfAssembly\MyWpfAssembly.csproj]
Done Building Project "C:\w\4\s\src\MyWpfAssembly\MyWpfAssembly.csproj" (default targets) -- FAILED.

```

Other customers have reported the same issue, ex.
https://social.msdn.microsoft.com/Forums/en-US/6360a1e3-a269-457c-a763-fb415b8bbf14/quotbecause-it-is-being-used-by-another-processquot-error-has-me-dead-in-the-water?forum=msbuild

---
#### Feedback Bot on 6/11/2021, 08:22 AM: 

<p>We have directed your feedback to the appropriate engineering team for further evaluation. The team will review the feedback and notify you about the next steps.</p>

#### Chaitanya Mangalagiri [MSFT] on 8/12/2021, 00:50 AM: 

(private comment, text removed)

---
### Original Solutions
(no solutions)

**Comment by timmi-on-rails** [comment]
Can you rule out, that a virus scanner or windows defender might be locking the file?
In my case it is working relyably on one machine, but not on the other.

**Comment by luy-ce** [comment]
> 

This failure occurs on our build agents, which as you can imagine, are set up to not have any external processes that may lock the file.

**Comment by timmi-on-rails** [comment]
Here comes a workaround that solves my (very similar) problem.

After the `MarkupCompilePass2` I am even unable to manually delete the DLL that the `CleanupTemporaryTargetAssembly` also fails to delete with a `Access to the path ... is denied.` error.
I was unable to determine the reason for the access denied error.
The command line tool https://docs.microsoft.com/en-us/sysinternals/downloads/handle showed no open file handles.
However I am running the build process inside a windows docker container, so maybe the host is locking the file.

The good thing is, I am still able to manually move the problematic DLL file.
So I added the following target to the project file.
Make sure to replace the file name.

```csproj
<Target Name="Hack" BeforeTargets="CleanupTemporaryTargetAssembly">
  <Exec Command="move Debug\AnyCPU\MyWpfAssembly.dll Debug\AnyCPU\MyWpfAssembly.txt" />
</Target>
```

Tell me, if this works for you as well?
There is still a chance that my problem is not connected to yours.

**Comment by adamlett** [comment]
I am also suffering from this issue, we have a large solution that contains many projects (C#, managed C++ and native C++ projects) and yet the only projects that suffer from this 'Unable to delete file' error are those that have just completed the 'MarkupCompilePass2' build task.

Initially we thought it could be an antivirus or similar locking the files, but this does not seem to be the case as I'd expect that to affect other projects too and not just the ones that undergo a second pass compile

Thanks for the workaround @timmi-on-rails, I'll see if it works for us too

**Comment by t-johnson** [comment]
@adamlett I've applied your workaround to avoid random build failures on our build host. Seems to be good for me!
I made a small change to support EXE and DLL outputs, as well as using $(IntermediateOutputPath) to simplify:
```
      <TemporaryAssemblyPath Condition="Exists('$(IntermediateOutputPath)$(AssemblyName).dll')">$(IntermediateOutputPath)$(AssemblyName).dll</TemporaryAssemblyPath>
      <TemporaryAssemblyPath Condition="Exists('$(IntermediateOutputPath)$(AssemblyName).exe')">$(IntermediateOutputPath)$(AssemblyName).exe</TemporaryAssemblyPath>
```

**Comment by Foitn** [comment]
Anyone found a solution?

**Comment by mattnorflus** [comment]
We (VSEng) are seeing this now in our build agents. We are setting up a new build environment to validate a VS build. This environment installs a test build of VS to an agent and tries to build the VS repo using the build tools of the test build. The intention is to verify that the VS product can build the VS repo. It is possible a change in the VS product may be surfacing this error but we have not been able to find a root cause. We believe this issue is related to our error at this time due to the similarities with what others have shared.

As other have mentioned above, we do not experience this problem on most of our agents, only this new build environment. The error occurs intermittently in different projects from build to build. The only common element of the failures is that `CleanupTemporaryTargetAssembly` fails to delete the binary after running `MarkupCompilePass2`. The above workarounds may not be simple for us to implement due to the size of the VS repo. It would be great to get more insight into the problem and maybe a fix? Our binlogs are large but can be found as artifacts for these builds:

https://dev.azure.com/devdiv/DevDiv/_build/results?buildId=8792392&view=logs&j=018464ba-1ca1-5561-2761-cd93078ac78a&t=3574efd9-071a-5092-469d-6ae99dc3b0cf
https://dev.azure.com/devdiv/DevDiv/_build/results?buildId=8790328&view=logs&j=018464ba-1ca1-5561-2761-cd93078ac78a&t=3574efd9-071a-5092-469d-6ae99dc3b0cf
https://dev.azure.com/devdiv/DevDiv/_build/results?buildId=8790291&view=logs&j=018464ba-1ca1-5561-2761-cd93078ac78a&t=3574efd9-071a-5092-469d-6ae99dc3b0cf

Feel free to ping me on teams if you are unable to view the builds and need to see the log for debugging.

**Comment by kyle-wilt-waters** [comment]
Could it be due to the implementation of MarkupCompilePass1 in PresentationBuildTasks? There is code here that creates an AppDomain, runs a compilation step, and then later unloads that AppDomain in the thread pool which effectively swallows any exceptions by default. So if the unload fails or takes a while and that AppDomain happens to include the dll of interest you'll hit this problem. This assumes AppDomain.Unload actually blocks until the domain is unloaded in the first place which isn't 100% clear to me based on the documentation and implementation I can see in 4.8.

There's another task named MarkupCompilePass2 which also creates an AppDomain for additional operations but it doesn't run the AppDomain.Unload asynchronously.

**Comment by khusmeno-MS** [comment]
I have the exact same problem. I added the workaround, I even increased the MaxWaitTimeMillis and timeBetweenTries, so it waits up to 1 hour, but still the build constantly fails. In the last 2 days, it succeeded only once! https://dev.azure.com/msazure/One/_build?definitionId=388978&_a=summary


![image](https://github.com/user-attachments/assets/04a4eeb3-63a2-4570-ad00-eed923a36eaf)

**Comment by bstordrup** [comment]
I also have exactly the same problem trying to build a WPF project in a docker container :-(

**Comment by sparrowt** [comment]
I'm hitting this same thing inside a windows docker container but only on certain servercore versinos:
- in `mcr.microsoft.com/windows/servercore:ltsc2025` it works fine ✅ 
- in `mcr.microsoft.com/windows/servercore:1809` (but with the exact same code & dependencies) it fails as follows: ❌
```
MarkupCompilePass2:
  MarkupCompilePass2 successfully generated BAML or source code files.
CleanupTemporaryTargetAssembly:
  Deleting file "path\to\X86\Release\my.exe".
C:\Windows\Microsoft.NET\Framework64\v4.0.30319\Microsoft.WinFX.targets(480,10): error MSB3061: Unable to delete file "path\to\X86\Release\my.exe". Access to the path 'path\to\X86\Release\my.exe' is denied. [path\to\my.csproj]
Done Building Project "path\to\my.csproj" (default targets) -- FAILED.

Build FAILED.
```
